lua/sim/units/components/TransportUnitComponent.lua

```diff
@@ -59,30 +65,78 @@ BaseTransport = ClassSimple {
     DestroyedOnTransport = function(self)
     end,
 
-    -- Detaches cargo from a dying unit
+    --- This function is called when a transport is killed. It kills all units inside the transport and
+    --- dispenses veterancy as appropriate.
+    --- For units that store their cargo externally, it caches the cargo for later impact.
     ---@param self BaseTransport | Unit
-    DetachCargo = function(self)
-        if self.Dead then return end -- Bail out early from overkill damage when already dead to avoid crashing
+    ---@param instigator Unit
+    ---@param damageType string
+    ---@param excessDamageRatio number
+    ---@param disperseVeterancy? boolean -- To prevent lower recursions from jumping the gun on veterancy
+    ---@return number? cargoMass -- The total mass value of the cargo
+    KillCargo = function(self, instigator, damageType, excessDamageRatio, disperseVeterancy)
+
+        -- If we're dead already, bail to avoid crashing
+        if self.Dead then return 0 end
 
         local cargo = self:GetCargo()
+
+        -- If we have storage slots, don't cache our cargo, because it's stored inside and
+        -- we'll never have to deal with it again
+        local cacheCargo = (self:GetBlueprint().Transport.StorageSlots and false) and true
+
+        if cacheCargo and not table.empty(cargo) then
+            LOG('Caching cargo')
+            self.killInstigator = instigator
+            self.killDamageType = damageType
+            self.killExcessDamageRatio = excessDamageRatio
+            self.cargoCache = {}
+        end
+
+        -- Establish a local variable for counting our cargo haul
+        local cargoMass = 0
+
         for _, unit in cargo do
-            if EntityCategoryContains(categories.TRANSPORTATION, unit) then -- Kill the contents of a transport in a transport, however that happened
-                for k, subUnit in unit:GetCargo() do
-                    subUnit:Kill()
-                end
+            -- Kill the contents of a transport in a transport, however that happened
+            -- now with recursion!
+            if EntityCategoryContains(categories.TRANSPORTATION, unit) then
+                cargoMass = cargoMass + unit:KillCargo(instigator,  damageType, excessDamageRatio, false)
             end
-            unit:DetachFrom()
+
+            -- cache the cargo so we can impact it later (if needed)
+            -- exception for command units, which explode immediately
+            if not EntityCategoryContains(categories.COMMAND, unit) and cacheCargo then
+                unit.killedInTransport = true
+                table.insert(self.cargoCache, unit)
+            end
```

Comment:
```suggestion
            if not EntityCategoryContains(categories.COMMAND, unit) then
                unit.killedInTransport = true
            end
            if cacheCargo then
                table.insert(self.cargoCache, unit)
            end
```
The two flags have independent purposes (triggering death effects from `Unit:Kill()` right now (on line 119) vs dealing with the units in a future time), so they should be treated separately.

Reply:
yep they gotta be separate

Reply:
They're actually dependent on each other. killedInTransport means "don't trigger OnKilled effects right now because we will deal with them later" and cacheCargo means "these are the units we are dealing with later". So they should be under the same if statement.
> triggering death effects from Unit:Kill() right now
That should be a third option as I described before, specific for carriers to "kill units now, but without effects".

Reply:
I knew there was a reason I put them together ;D

But yes you're definitely right about the third option, I'll assemble it.